File: python/102-Binary-Tree-Level-Order-Traversal.py
# Definition for a binary tree node.
from ast import List
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# region Recursive BFS
# Space O(n)
# Time O(n)
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        retVal = []
        self.traverse(root, retVal, 0)
        return retVal

    def traverse(self, root: Optional[TreeNode], retVal: List[List[int]], level: int) -> None:
        if root is None:
            return
        while len(retVal) <= level:
            retVal.append([])
        retVal[level].append(root.val)
        self.traverse(root.left, retVal, level + 1)
        self.traverse(root.right, retVal, level + 1)


# endregion

# region Iterative BFS
# Time O(n)
# Space O(n)


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        retVal = []

        if not root:
            return retVal

        q = deque()
        q.append(root)
        level = 0
        while q:
            size = len(q)

            while size > 0:
                if len(retVal) <= level:
                    retVal.append([])

                cur = q.popleft()
                if cur:
                    retVal[level].append(cur.val)
                    if cur.left:
                        q.append(cur.left)
                    if cur.right:
                        q.append(cur.right)
                size -= 1
            level += 1
        return retVal


# endregion

File: c#/102-Binary-Tree-Level-Order-Traversal.cs
#region Breadth First Search
public class Solution {
    public List<List<int>> LevelOrder(TreeNode root) {
        var retVal = new List<List<int>>();
        if (root == null) return retVal;

        var q = new Queue<TreeNode>();
        q.Enqueue(root);

        while (q.Count > 0) {
            var level = new List<int>();
            int levelCount = q.Count;

            for (int i = 0; i < levelCount; i++) {
                var cur = q.Dequeue();

                if (cur.left != null) q.Enqueue(cur.left);
                if (cur.right != null) q.Enqueue(cur.right);

                level.Add(cur.val);
            }

            retVal.Add(level);
        }

        return retVal;
    }
}

#endregion

#region Depth First Search
public class Solution {
    public List<List<int>> res = new List<List<int>>();

    public List<List<int>> LevelOrder(TreeNode root) {
        dfs(root, 0);
        return res;
    }

    private void dfs(TreeNode node, int depth) {
        if (node == null) {
            return;
        }

        if (res.Count == depth) {
            res.Add(new List<int>());
        }

        res[depth].Add(node.val);

        dfs(node.left, depth + 1);
        dfs(node.right, depth + 1);
    }
}

#endregion